### PR TITLE
JBPM-9273 System error occurred while adding deployment unit

### PR DESCRIPTION
**Pull request by bxf12315**
**Thank you for submitting this pull request**

**JIRA**: [JBPM-9273](https://issues.redhat.com/browse/JBPM-9273)

**Referenced Pull Requests**:
* https://github.com/kiegroup/droolsjbpm-integration/pull/2367
* link 2
* link 3 etc.

**Business Central**: [WAR file](https://donwnload.here/something)

**VS Code**: [plugin](https://donwnload.here/something)

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* Retest PR: <b>jenkins retest this</b>
* A full downstream build: <b>jenkins do fdb</b>
* A compile downstream build: <b>jenkins do cdb</b>
* A full production downstream build: <b>jenkins do product fdb</b>
* An upstream build: <b>jenkins do upstream</b>
</details>

### CODE REVIEW

**Comment by bxf12315** [comment]
dependent on https://github.com/kiegroup/droolsjbpm-integration/pull/2367

**Comment by prashuts** [comment]
jenkins do fdb

**Review by nmirasch — APPROVED** [review]

File: kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-client/src/main/java/org/kie/workbench/common/screens/server/management/client/container/ContainerPresenter.java [hunk 1]
```diff
@@ -143,13 +147,28 @@ public void loadContainers(@Observes final ContainerSpecData content) {
...
+        containers.forEach(container -> {
+            if (KieContainerStatus.FAILED == container.getStatus() || container.getResolvedReleasedId() == null) {
+                container.setResolvedReleasedId(containerSpec.getReleasedId());
+                container.addMessage(new Message(Severity.ERROR, Collections.emptyList()));
+            }
+        });
+        Optional<Container> optionalContainer = containers.stream().filter(container -> KieContainerStatus.FAILED != container.getStatus()).findFirst();
+        if (!optionalContainer.isPresent() && containers.size() > 0) {
+            containerSpec.setStatus(KieContainerStatus.FAILED);
+        }
```
Comment (sutaakar) [inline kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-client/src/main/java/org/kie/workbench/common/screens/server/management/client/container/ContainerPresenter.java:168]: What is the purpose of this code?

Comment (bxf12315) [reply kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-client/src/main/java/org/kie/workbench/common/screens/server/management/client/container/ContainerPresenter.java:168]: if all containers is failed of ContinaerSpec,will set the ContinaerSpec status is failed.

Comment (bxf12315) [reply kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-client/src/main/java/org/kie/workbench/common/screens/server/management/client/container/ContainerPresenter.java:168]: ContainerSpec failed logic in here
https://github.com/kiegroup/kie-wb-common/pull/3539/files#diff-ba633635b053c1407bae080eda756c262a60387af1c9b930d93b78fe1d98feaeR266

Comment (sutaakar) [reply kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-client/src/main/java/org/kie/workbench/common/screens/server/management/client/container/ContainerPresenter.java:168]: Can you please move the code into separate method? It doesn't have any relation with resetting release id for failed containers.

File: kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-client/src/main/java/org/kie/workbench/common/screens/server/management/client/container/ContainerPresenter.java [hunk 2]
```diff
@@ -143,13 +147,28 @@ public void loadContainers(@Observes final ContainerSpecData content) {
...
     }
 
+    private void resetReleaseIdForFailedContainers(Collection<Container> containers, ContainerSpec containerSpec) {
+        containers.forEach(container -> {
+            if (KieContainerStatus.FAILED == container.getStatus() || container.getResolvedReleasedId() == null) {
+                container.setResolvedReleasedId(containerSpec.getReleasedId());
+                container.addMessage(new Message(Severity.ERROR, Collections.emptyList()));
+            }
+        });
+        Optional<Container> optionalContainer = containers.stream().filter(container -> KieContainerStatus.FAILED != container.getStatus()).findFirst();
```
Comment (sutaakar) [inline kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-client/src/main/java/org/kie/workbench/common/screens/server/management/client/container/ContainerPresenter.java:165]: The condition is not well readable, also `optionalContainer` name doesn't tell much about its content.
Can you for example rather use `allMatch` predicate with variable name like `allContainersAreFailed`?

File: kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-client/src/main/java/org/kie/workbench/common/screens/server/management/client/container/ContainerPresenter.java (hunk 1 above)
Comment (bxf12315) [reply kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-client/src/main/java/org/kie/workbench/common/screens/server/management/client/container/ContainerPresenter.java]: Done

File: kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-client/src/main/java/org/kie/workbench/common/screens/server/management/client/container/ContainerPresenter.java (hunk 2 above)
Comment (bxf12315) [reply kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-client/src/main/java/org/kie/workbench/common/screens/server/management/client/container/ContainerPresenter.java:171]: Done

**Comment by bxf12315** [comment]
jenkins do fdb

**Comment by prashuts** [comment]
@bxf12315  As per https://issues.redhat.com/browse/JBPM-9273  , 'STOP' and 'Deactivate' buttons are supposed to be disabled while adding deployment unit with error warning alert.(Ref - https://issues.redhat.com/secure/attachment/12495863/20200922180910842.mp4)..
Is this still a valid flow ? 
Please confirm on this.

File: kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-client/src/main/java/org/kie/workbench/common/screens/server/management/client/container/ContainerPresenter.java (hunk 2 above)
Comment (sutaakar) [reply kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-client/src/main/java/org/kie/workbench/common/screens/server/management/client/container/ContainerPresenter.java:171]: Actually it is not done, the comment is still valid.

Comment (bxf12315) [reply kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-client/src/main/java/org/kie/workbench/common/screens/server/management/client/container/ContainerPresenter.java:171]: I refactor them to private method, not just a re-name.

Comment (sutaakar) [reply kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-client/src/main/java/org/kie/workbench/common/screens/server/management/client/container/ContainerPresenter.java:171]: This comment is about this line:
` Optional<Container> optionalContainer = containers.stream().filter(container -> KieContainerStatus.FAILED != container.getStatus()).findFirst();`
It is relatively difficult to read it, takes time to actually understand it for casual reader.

Comment (bxf12315) [reply kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-client/src/main/java/org/kie/workbench/common/screens/server/management/client/container/ContainerPresenter.java]: ok

**Comment by bxf12315** [comment]
> issues.redhat.com/secure/attachment/12495863/20200922180910842.mp4

For my side, yes

**Comment by prashuts** [comment]
> > issues.redhat.com/secure/attachment/12495863/20200922180910842.mp4
> 
> For my side, yes'

Its not working as shown in video , please check on it. 'Stop' and 'Deactivate' buttons are still enabled after error alert.

**Comment by bxf12315** [comment]
> > > issues.redhat.com/secure/attachment/12495863/20200922180910842.mp4
> > 
> > 
> > For my side, yes'
> 
> Its not working as shown in video , please check on it. 'Stop' and 'Deactivate' buttons are still enabled after error alert.

Please wait a few more seconds, I found that this event was not notified immediately.

**Comment by prashuts** [comment]
jenkins do fdb

**Comment by bxf12315** [comment]
jenkins retest this

**Comment by prashuts** [comment]
> > > > issues.redhat.com/secure/attachment/12495863/20200922180910842.mp4
> > > 
> > > 
> > > For my side, yes'
> > 
> > 
> > Its not working as shown in video , please check on it. 'Stop' and 'Deactivate' buttons are still enabled after error alert.
> 
> Please wait a few more seconds, I found that this event was not notified immediately.

No luck even after waiting.

**Comment by bxf12315** [comment]
Jenkins run fdb

**Comment by bxf12315** [comment]
Jenkins run fdb

**Comment by prashuts** [comment]
jenkins do fdb

**Comment by bxf12315** [comment]
jenkins do fdb

**Comment by bxf12315** [comment]
jenkins do fdb

**Comment by bxf12315** [comment]
jenkins retest this

**Comment by bxf12315** [comment]
jenkins retest this

**Review by sutaakar — APPROVED** [review]
Looks good to me.
Works as expected for me (when deploying non existing container then error message is shown and it is possible just to remove the container)

**Review by prashuts — APPROVED** [review]
LGTM

**Pull request merged**